File: BetterChannelIcons/src/main/java/xyz/wingio/plugins/betterchannelicons/AddChannelSheet.java
package xyz.wingio.plugins.betterchannelicons;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.*;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RectShape;
import android.graphics.Bitmap.Config;
import android.graphics.PorterDuff.Mode;
import android.content.res.Resources;
import android.content.res.AssetManager;
import android.net.Uri;
import android.text.*;
import android.text.style.ClickableSpan;
import android.util.Base64;
import android.view.*;
import android.widget.*;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.*;
import androidx.core.content.res.ResourcesCompat;

import xyz.wingio.plugins.BetterChannelIcons;
import com.aliucord.Utils;
import com.aliucord.Logger;
import com.aliucord.PluginManager;
import com.aliucord.api.SettingsAPI;
import com.aliucord.fragments.SettingsPage;
import com.aliucord.utils.ReflectUtils;
import com.aliucord.widgets.BottomSheet;
import com.discord.widgets.user.Badge;
import com.aliucord.views.TextInput;
import com.aliucord.views.Button;
import com.discord.app.AppBottomSheet;
import com.discord.app.AppFragment;
import com.discord.views.CheckedSetting;
import com.discord.utilities.color.ColorCompat;
import com.lytefast.flexinput.R;

import kotlin.Unit;
import java.io.*;
import java.util.*;

public class AddChannelSheet extends BottomSheet {
    private SettingsPage page;
    private SettingsAPI settings;
    private String currentIcon;
    private ImageView icon;

    public AddChannelSheet(SettingsPage page, SettingsAPI settings) {
        this.page = page;
        this.settings = settings;
    }

    @Override
    public void onViewCreated(View view, Bundle bundle) {
        super.onViewCreated(view, bundle);
        int p = Utils.dpToPx(16);
        Map<String, String> iconSets = settings.getObject("icons", new HashMap<>(), BetterChannelIcons.iconStoreType);
        
        setPadding(p);
        Context ctx = requireContext();
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        params.setMargins(0, 0, 0, p);


        LinearLayout iconLayout = new LinearLayout(ctx);
        iconLayout.setOrientation(LinearLayout.HORIZONTAL);
        iconLayout.setGravity(Gravity.CENTER);
        iconLayout.setLayoutParams(params);

        TextInput channelName = new TextInput(ctx);
        channelName.setHint("Channel name");
        LinearLayout.LayoutParams iconLParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        channelName.setLayoutParams(iconLParams);
        channelName.getEditText().setText("");

        var resources = ctx.getResources();
        Drawable badge = ResourcesCompat.getDrawable(resources, R.d.ic_open_in_new_grey_24dp, null).mutate();
        badge.setTint(ColorCompat.getThemedColor(ctx, R.b.colorInteractiveNormal));

        icon = new ImageView(ctx);
        icon.setImageDrawable(badge);
        var iconParams = new LinearLayout.LayoutParams(Utils.dpToPx(32), Utils.dpToPx(32));
        iconParams.setMargins(p / 2, 0, p, 0);
        icon.setLayoutParams(iconParams);
        icon.setScaleType(ImageView.ScaleType.FIT_CENTER);
        icon.setOnClickListener(v -> {
            new IconListSheet(page, settings, this).show(getFragmentManager(), "icon_list");
        });
        
        Button add = new Button(ctx);
        add.setText("Add");

        iconLayout.addView(icon);
        iconLayout.addView(channelName);

        add.setOnClickListener(v -> {
            String name = channelName.getEditText().getText().toString();
            if (name.isEmpty()) {
                Toast.makeText(ctx, "Please enter a channel name", Toast.LENGTH_SHORT).show();
            } else {
                try {
                    var ic = ContextCompat.getDrawable(ctx, Utils.getResId(currentIcon, "drawable"));
                    var chName = name.toLowerCase();
                    if(!iconSets.containsKey(chName)){
                        iconSets.put(chName, currentIcon);
                        settings.setObject("icons", iconSets);
                        Toast.makeText(ctx, "Added icon", Toast.LENGTH_SHORT).show();
                    }
                    page.reRender();
                    dismiss();
                } catch (Throwable e) {
                    Logger logger = new Logger("BCI");
                    logger.error("Error setting icon", e);
                    Toast.makeText(ctx, "Drawable not found", Toast.LENGTH_SHORT).show();
                }
            }
        });

        addView(iconLayout);
        addView(add);
    }

    public void setCurrentIcon(String icon) {
        this.currentIcon = icon;
        if(this.icon != null){
            var ic = ContextCompat.getDrawable(requireContext(), Utils.getResId(icon, "drawable")).mutate();
            ic.setTint(ColorCompat.getThemedColor(requireContext(), R.b.colorInteractiveNormal));
            this.icon.setImageDrawable(ic);
        }
    }

    private CheckedSetting createSwitch(Context context, SettingsAPI sets, String key, String label, CharSequence subtext, boolean defaultValue, boolean reRender) {
        CheckedSetting cs = Utils.createCheckedSetting(context, CheckedSetting.ViewType.SWITCH, label, subtext);
        cs.setChecked(sets.getBool(key, defaultValue));
        cs.setOnCheckedListener(c -> {
            sets.setBool(key, c);
        });
        return cs;
    }
}

File: BetterChannelIcons/src/main/java/xyz/wingio/plugins/betterchannelicons/AddChannelSheet.kt
@file:Suppress("MISSING_DEPENDENCY_CLASS", "MISSING_DEPENDENCY_SUPERCLASS")

package xyz.wingio.plugins.betterchannelicons

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.Toast

import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat

import com.aliucord.Utils.getResId
import com.aliucord.Utils.tintToTheme
import com.aliucord.api.SettingsAPI
import com.aliucord.fragments.SettingsPage
import com.aliucord.utils.DimenUtils.dp
import com.aliucord.views.Button
import com.aliucord.views.TextInput
import com.aliucord.widgets.BottomSheet

import com.lytefast.flexinput.R

import xyz.wingio.plugins.BetterChannelIcons

import java.util.Locale

class AddChannelSheet : BottomSheet {

    private var page: SettingsPage? = null
    private var settings: SettingsAPI
    private var currentIcon: String? = null
    private var icon: ImageView? = null
    private var channelId: Long? = null

    constructor(page: SettingsPage?, settings: SettingsAPI) {
        this.page = page
        this.settings = settings
    }

    constructor(channelId: Long?, settings: SettingsAPI) {
        this.channelId = channelId
        this.settings = settings
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, bundle: Bundle?) {
        super.onViewCreated(view, bundle)
        val ctx = requireContext()
        val resources = ctx.resources
        val badge = tintToTheme(ResourcesCompat.getDrawable(resources, R.e.ic_open_in_new_grey_24dp, null)!!.mutate())
        val iconSets: MutableMap<String, String?> =
            settings.getObject("icons", HashMap(), Utils.iconStoreType)

        setPadding(16.dp)

        val iconLayout = LinearLayout(ctx).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).apply { setMargins(0, 0, 0, 16.dp) }
        }

        val channelNameInput = TextInput(ctx).apply {
            editText.setText(if (channelId == null || channelId == 0L) "" else "id:$channelId")
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )

            setHint("Channel name")
        }

        icon = ImageView(ctx).apply {
            scaleType = ImageView.ScaleType.FIT_CENTER
            layoutParams = LinearLayout.LayoutParams(32.dp, 32.dp).apply {
                setMargins(8.dp, 0, 8.dp, 0)
            }

            setImageDrawable(badge)
            setOnClickListener {
                IconListSheet(
                    settings = settings,
                    addChannelSheet = this@AddChannelSheet
                ).show(parentFragmentManager, "icon_list")
            }
        }

        val add = Button(ctx).apply {
            text = "Add"

            setOnClickListener {
                val name = channelNameInput.editText.text.toString()

                if (name.isEmpty()) {
                    Toast.makeText(ctx, "Please enter a channel name", Toast.LENGTH_SHORT).show()
                } else {
                    try {
                        val channelName = name.lowercase(Locale.getDefault())

                        // This is just done to verify if the provided icon actually exists
                        ContextCompat.getDrawable(
                            /* context = */ ctx,
                            /* id = */ getResId(
                                name = currentIcon!!,
                                type = "drawable"
                            )
                        )

                        iconSets[channelName] = currentIcon
                        settings.setObject("icons", iconSets)

                        Toast.makeText(ctx, "Added icon", Toast.LENGTH_SHORT).show()

                        if (page != null) page!!.reRender()
                        dismiss()
                    } catch (e: Throwable) {
                        BetterChannelIcons.logger.error("Error setting icon", e)
                        Toast.makeText(ctx, "Drawable not found", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }

        iconLayout.addView(icon)
        iconLayout.addView(channelNameInput)

        addView(iconLayout)
        addView(add)
    }

    fun setCurrentIcon(icon: String?) {
        currentIcon = icon

        if (this.icon != null) {
            val ic = ContextCompat.getDrawable(
                /* context = */ requireContext(),
                /* id = */ getResId(
                    name = icon!!,
                    type = "drawable"
                )
            )!!.mutate()

            this.icon!!.setImageDrawable(tintToTheme(ic))
        }
    }

}